[java] java.lang.IllegalArgumentException: <?> cannot be a wildcard bound

<!-- Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core]. -->

**Affects PMD Version:** 

7.4.0

**Description:**

PMD throws an IllegalArgumentException while processing valid Java code.

**Exception Stacktrace:**

```
2024-08-13T11:59:02.7818012Z [INFO] --- pmd:3.24.0:pmd (pmd) @ nms ---
2024-08-13T11:59:03.0987733Z [INFO] PMD version: 7.4.0
2024-08-13T11:59:11.1328793Z [WARNING] Exception applying rule LiteralsFirstInComparisons on file /home/runner/work/song-nms/song-nms/nms/src/main/java/com/sg/song/nms/security/impl/SecurityCdiExtension.java, continuing with next rule
2024-08-13T11:59:11.1340402Z org.apache.commons.lang3.exception.ContextedRuntimeException: java.lang.IllegalArgumentException: <?> cannot be a wildcard bound
2024-08-13T11:59:11.1342480Z Exception Context:
2024-08-13T11:59:11.1343299Z 	[1:Resolving type of=!debug only! [VariableId:124:7]method]
2024-08-13T11:59:11.1344268Z 	[2:Resolving type of=!debug only! [VariableAccess:134:7]method]
2024-08-13T11:59:11.1736408Z 	[3:Resolving type of=!debug only! [MethodCall:134:7]method.getAnnotated()]
2024-08-13T11:59:11.1738342Z 	[4:Resolving type of=!debug only! [MethodCall:134:7]method.getAnnotated().getParameters()]
2024-08-13T11:59:11.1740129Z 	[5:Resolving type of=!debug only! [MethodCall:134:7]method.getAnnotated().getParameters().stream()]
2024-08-13T11:59:11.1742788Z 	[6:Resolving type of=!debug only! [MethodCall:134:7]method.getAnnotated().getParameters().stream().noneMatch(p -> p.getBaseType().equals(Session.class))]
2024-08-13T11:59:11.1745009Z 	[7:Resolving type of=!debug only! [MethodCall:134:69]p.getBaseType().equals(Session.class)]
2024-08-13T11:59:11.1746755Z 	[8:Rule applied on node=!debug only! [MethodCall:134:69]p.getBaseType().equals(Session.class)]
2024-08-13T11:59:11.1748427Z ---------------------------------
2024-08-13T11:59:11.1749772Z     at net.sourceforge.pmd.util.AssertionUtil.contexted (AssertionUtil.java:236)
2024-08-13T11:59:11.1751669Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:54)
2024-08-13T11:59:11.1753559Z     at net.sourceforge.pmd.lang.java.ast.ASTVariableId.getTypeMirror (ASTVariableId.java:46)
2024-08-13T11:59:11.1755434Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:39)
2024-08-13T11:59:11.1757302Z     at net.sourceforge.pmd.lang.java.ast.ASTVariableId.getTypeMirror (ASTVariableId.java:341)
2024-08-13T11:59:11.1759282Z     at net.sourceforge.pmd.lang.java.symbols.internal.ast.AstLocalVarSym.getTypeMirror (AstLocalVarSym.java:40)
2024-08-13T11:59:11.1761226Z     at net.sourceforge.pmd.lang.java.types.JVariableSig.getTypeMirror (JVariableSig.java:93)
2024-08-13T11:59:11.1763104Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:616)
2024-08-13T11:59:11.1765062Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.1766992Z     at net.sourceforge.pmd.lang.java.ast.ASTVariableAccess.acceptVisitor (ASTVariableAccess.java:76)
2024-08-13T11:59:11.1768940Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.1770891Z     at net.sourceforge.pmd.lang.java.ast.ASTVariableAccess.getTypeMirror (ASTVariableAccess.java:27)
2024-08-13T11:59:11.1773175Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:66)
2024-08-13T11:59:11.1776009Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:26)
2024-08-13T11:59:11.1778240Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.computeCompileTimeDecl (Infer.java:303)
2024-08-13T11:59:11.1780224Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.getCompileTimeDecl (Infer.java:281)
2024-08-13T11:59:11.1782215Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.goToInvocationWithFallback (Infer.java:214)
2024-08-13T11:59:11.1784734Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.inferInvocationRecursively (Infer.java:176)
2024-08-13T11:59:11.1786887Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.inferInvocation (PolyResolution.java:263)
2024-08-13T11:59:11.1789058Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeOtherCtx (PolyResolution.java:135)
2024-08-13T11:59:11.1791626Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:125)
2024-08-13T11:59:11.1793840Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.1795917Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.1797901Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.1799778Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.1801659Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.1803512Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.1805751Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:66)
2024-08-13T11:59:11.1808443Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:26)
2024-08-13T11:59:11.1810752Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.computeCompileTimeDecl (Infer.java:303)
2024-08-13T11:59:11.1812710Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.getCompileTimeDecl (Infer.java:281)
2024-08-13T11:59:11.1815090Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.goToInvocationWithFallback (Infer.java:214)
2024-08-13T11:59:11.1817027Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.inferInvocationRecursively (Infer.java:176)
2024-08-13T11:59:11.1818986Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.inferInvocation (PolyResolution.java:263)
2024-08-13T11:59:11.1821002Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeOtherCtx (PolyResolution.java:135)
2024-08-13T11:59:11.1823029Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:125)
2024-08-13T11:59:11.1825027Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.1826986Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.1828864Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.1830641Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.1832392Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.1834097Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.1840869Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:66)
2024-08-13T11:59:11.1843418Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:26)
2024-08-13T11:59:11.1850643Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.computeCompileTimeDecl (Infer.java:303)
2024-08-13T11:59:11.1852551Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.getCompileTimeDecl (Infer.java:281)
2024-08-13T11:59:11.1863191Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.goToInvocationWithFallback (Infer.java:214)
2024-08-13T11:59:11.1865379Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.inferInvocationRecursively (Infer.java:176)
2024-08-13T11:59:11.1867521Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.inferInvocation (PolyResolution.java:263)
2024-08-13T11:59:11.1869606Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeOtherCtx (PolyResolution.java:135)
2024-08-13T11:59:11.1871998Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:125)
2024-08-13T11:59:11.1874077Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.1876039Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.1878072Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.1902791Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.1904811Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.1918528Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.1920776Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:66)
2024-08-13T11:59:11.1923440Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:26)
2024-08-13T11:59:11.1925780Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.computeCompileTimeDecl (Infer.java:303)
2024-08-13T11:59:11.1927740Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.getCompileTimeDecl (Infer.java:281)
2024-08-13T11:59:11.1929753Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.goToInvocationWithFallback (Infer.java:214)
2024-08-13T11:59:11.1931788Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.inferInvocationRecursively (Infer.java:176)
2024-08-13T11:59:11.1934111Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.inferInvocation (PolyResolution.java:263)
2024-08-13T11:59:11.1936329Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeOtherCtx (PolyResolution.java:135)
2024-08-13T11:59:11.1938493Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:125)
2024-08-13T11:59:11.1940762Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.1942837Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.1944901Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.1946833Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.1948656Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.1950468Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.1952315Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:39)
2024-08-13T11:59:11.1954210Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.1956177Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeInvocationCtx (PolyResolution.java:217)
2024-08-13T11:59:11.1958302Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:122)
2024-08-13T11:59:11.1960836Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.1962898Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.1964889Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.1966949Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.1968836Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.1970714Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.1972613Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:39)
2024-08-13T11:59:11.1974649Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.1976580Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.forceTypeResolution (AbstractJavaTypeNode.java:29)
2024-08-13T11:59:11.1978746Z     at net.sourceforge.pmd.lang.java.ast.AbstractInvocationExpr.getOverloadSelectionInfo (AbstractInvocationExpr.java:27)
2024-08-13T11:59:11.1980892Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getOverloadSelectionInfo (ASTMethodCall.java:22)
2024-08-13T11:59:11.1983549Z     at net.sourceforge.pmd.lang.java.rule.bestpractices.LiteralsFirstInComparisonsRule.isEqualsObjectAndNotAnOverload (LiteralsFirstInComparisonsRule.java:52)
2024-08-13T11:59:11.1986546Z     at net.sourceforge.pmd.lang.java.rule.bestpractices.LiteralsFirstInComparisonsRule.visit (LiteralsFirstInComparisonsRule.java:41)
2024-08-13T11:59:11.1988630Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.1990383Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.acceptVisitor (AbstractJavaNode.java:38)
2024-08-13T11:59:11.1992095Z     at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply (AbstractJavaRule.java:30)
2024-08-13T11:59:11.1993592Z     at net.sourceforge.pmd.lang.rule.RuleReference.apply (RuleReference.java:415)
2024-08-13T11:59:11.1995251Z     at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:77)
2024-08-13T11:59:11.1997013Z     at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:57)
2024-08-13T11:59:11.1998619Z     at net.sourceforge.pmd.lang.rule.internal.RuleSets.apply (RuleSets.java:155)
2024-08-13T11:59:11.2000164Z     at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
2024-08-13T11:59:11.2001641Z     at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
2024-08-13T11:59:11.2003022Z     at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
2024-08-13T11:59:11.2004283Z     at java.util.concurrent.FutureTask.run (FutureTask.java:264)
2024-08-13T11:59:11.2154443Z     at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
2024-08-13T11:59:11.2156109Z     at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
2024-08-13T11:59:11.2157343Z     at java.lang.Thread.run (Thread.java:840)
2024-08-13T11:59:11.2158445Z Caused by: java.lang.IllegalArgumentException: <?> cannot be a wildcard bound
2024-08-13T11:59:11.2159943Z     at net.sourceforge.pmd.lang.java.types.TypeSystem.wildcard (TypeSystem.java:640)
2024-08-13T11:59:11.2161616Z     at net.sourceforge.pmd.lang.java.types.TypeOps$ProjectionVisitor.visitClass (TypeOps.java:1183)
2024-08-13T11:59:11.2177341Z     at net.sourceforge.pmd.lang.java.types.TypeOps$ProjectionVisitor.visitClass (TypeOps.java:1123)
2024-08-13T11:59:11.2179057Z     at net.sourceforge.pmd.lang.java.types.JClassType.acceptVisitor (JClassType.java:312)
2024-08-13T11:59:11.2180632Z     at net.sourceforge.pmd.lang.java.types.TypeOps.projectUpwards (TypeOps.java:1006)
2024-08-13T11:59:11.2182798Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:232)
2024-08-13T11:59:11.2184778Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.2186567Z     at net.sourceforge.pmd.lang.java.ast.ASTVariableId.acceptVisitor (ASTVariableId.java:57)
2024-08-13T11:59:11.2188616Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.2190381Z     at net.sourceforge.pmd.lang.java.ast.ASTVariableId.getTypeMirror (ASTVariableId.java:46)
2024-08-13T11:59:11.2192082Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:39)
2024-08-13T11:59:11.2193841Z     at net.sourceforge.pmd.lang.java.ast.ASTVariableId.getTypeMirror (ASTVariableId.java:341)
2024-08-13T11:59:11.2195738Z     at net.sourceforge.pmd.lang.java.symbols.internal.ast.AstLocalVarSym.getTypeMirror (AstLocalVarSym.java:40)
2024-08-13T11:59:11.2197539Z     at net.sourceforge.pmd.lang.java.types.JVariableSig.getTypeMirror (JVariableSig.java:93)
2024-08-13T11:59:11.2415884Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:616)
2024-08-13T11:59:11.2417842Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.2419792Z     at net.sourceforge.pmd.lang.java.ast.ASTVariableAccess.acceptVisitor (ASTVariableAccess.java:76)
2024-08-13T11:59:11.2421731Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.2423660Z     at net.sourceforge.pmd.lang.java.ast.ASTVariableAccess.getTypeMirror (ASTVariableAccess.java:27)
2024-08-13T11:59:11.2425920Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:66)
2024-08-13T11:59:11.2428602Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:26)
2024-08-13T11:59:11.2430839Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.computeCompileTimeDecl (Infer.java:303)
2024-08-13T11:59:11.2432717Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.getCompileTimeDecl (Infer.java:281)
2024-08-13T11:59:11.2434606Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.goToInvocationWithFallback (Infer.java:214)
2024-08-13T11:59:11.2436676Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.inferInvocationRecursively (Infer.java:176)
2024-08-13T11:59:11.2438790Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.inferInvocation (PolyResolution.java:263)
2024-08-13T11:59:11.2440953Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeOtherCtx (PolyResolution.java:135)
2024-08-13T11:59:11.2443139Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:125)
2024-08-13T11:59:11.2445231Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.2447321Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.2449238Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.2451078Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.2452942Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.2467341Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.2469540Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:66)
2024-08-13T11:59:11.2472588Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:26)
2024-08-13T11:59:11.2474843Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.computeCompileTimeDecl (Infer.java:303)
2024-08-13T11:59:11.2476713Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.getCompileTimeDecl (Infer.java:281)
2024-08-13T11:59:11.2478940Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.goToInvocationWithFallback (Infer.java:214)
2024-08-13T11:59:11.2480934Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.inferInvocationRecursively (Infer.java:176)
2024-08-13T11:59:11.2483046Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.inferInvocation (PolyResolution.java:263)
2024-08-13T11:59:11.2485156Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeOtherCtx (PolyResolution.java:135)
2024-08-13T11:59:11.2487319Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:125)
2024-08-13T11:59:11.2489482Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.2491535Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.2493525Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.2548046Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.2549860Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.2551642Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.2553727Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:66)
2024-08-13T11:59:11.2556252Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:26)
2024-08-13T11:59:11.2558440Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.computeCompileTimeDecl (Infer.java:303)
2024-08-13T11:59:11.2623592Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.getCompileTimeDecl (Infer.java:281)
2024-08-13T11:59:11.2625699Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.goToInvocationWithFallback (Infer.java:214)
2024-08-13T11:59:11.2627772Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.inferInvocationRecursively (Infer.java:176)
2024-08-13T11:59:11.2629830Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.inferInvocation (PolyResolution.java:263)
2024-08-13T11:59:11.2631918Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeOtherCtx (PolyResolution.java:135)
2024-08-13T11:59:11.2634038Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:125)
2024-08-13T11:59:11.2636121Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.2638127Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.2640063Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.2641832Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.2643615Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.2645403Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.2647475Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:66)
2024-08-13T11:59:11.2650336Z     at net.sourceforge.pmd.lang.java.types.internal.infer.ast.MethodInvocMirror.getAccessibleCandidates (MethodInvocMirror.java:26)
2024-08-13T11:59:11.2652555Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.computeCompileTimeDecl (Infer.java:303)
2024-08-13T11:59:11.2654634Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.getCompileTimeDecl (Infer.java:281)
2024-08-13T11:59:11.2656872Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.goToInvocationWithFallback (Infer.java:214)
2024-08-13T11:59:11.2658907Z     at net.sourceforge.pmd.lang.java.types.internal.infer.Infer.inferInvocationRecursively (Infer.java:176)
2024-08-13T11:59:11.2660971Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.inferInvocation (PolyResolution.java:263)
2024-08-13T11:59:11.2663085Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeOtherCtx (PolyResolution.java:135)
2024-08-13T11:59:11.2665204Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:125)
2024-08-13T11:59:11.2667283Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.2669295Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.2671269Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.2673079Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.2674926Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.2676738Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.2678547Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:39)
2024-08-13T11:59:11.2680341Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.2682331Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.polyTypeInvocationCtx (PolyResolution.java:217)
2024-08-13T11:59:11.2684487Z     at net.sourceforge.pmd.lang.java.types.ast.internal.PolyResolution.computePolyType (PolyResolution.java:122)
2024-08-13T11:59:11.2686573Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.handlePoly (LazyTypeResolver.java:351)
2024-08-13T11:59:11.2688604Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:356)
2024-08-13T11:59:11.2690566Z     at net.sourceforge.pmd.lang.java.types.ast.internal.LazyTypeResolver.visit (LazyTypeResolver.java:102)
2024-08-13T11:59:11.2692364Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.2930145Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:51)
2024-08-13T11:59:11.2932165Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.2934241Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.getTypeMirror (AbstractJavaTypeNode.java:39)
2024-08-13T11:59:11.2936078Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getTypeMirror (ASTMethodCall.java:22)
2024-08-13T11:59:11.2937872Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaTypeNode.forceTypeResolution (AbstractJavaTypeNode.java:29)
2024-08-13T11:59:11.2940016Z     at net.sourceforge.pmd.lang.java.ast.AbstractInvocationExpr.getOverloadSelectionInfo (AbstractInvocationExpr.java:27)
2024-08-13T11:59:11.2942075Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.getOverloadSelectionInfo (ASTMethodCall.java:22)
2024-08-13T11:59:11.2945016Z     at net.sourceforge.pmd.lang.java.rule.bestpractices.LiteralsFirstInComparisonsRule.isEqualsObjectAndNotAnOverload (LiteralsFirstInComparisonsRule.java:52)
2024-08-13T11:59:11.2947801Z     at net.sourceforge.pmd.lang.java.rule.bestpractices.LiteralsFirstInComparisonsRule.visit (LiteralsFirstInComparisonsRule.java:41)
2024-08-13T11:59:11.2951950Z     at net.sourceforge.pmd.lang.java.ast.ASTMethodCall.acceptVisitor (ASTMethodCall.java:71)
2024-08-13T11:59:11.2953624Z     at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.acceptVisitor (AbstractJavaNode.java:38)
2024-08-13T11:59:11.2955644Z     at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply (AbstractJavaRule.java:30)
2024-08-13T11:59:11.2957104Z     at net.sourceforge.pmd.lang.rule.RuleReference.apply (RuleReference.java:415)
2024-08-13T11:59:11.2958709Z     at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:77)
2024-08-13T11:59:11.2960366Z     at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:57)
2024-08-13T11:59:11.2961899Z     at net.sourceforge.pmd.lang.rule.internal.RuleSets.apply (RuleSets.java:155)
2024-08-13T11:59:11.2963300Z     at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
2024-08-13T11:59:11.2964668Z     at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
2024-08-13T11:59:11.2965917Z     at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
2024-08-13T11:59:11.2967133Z     at java.util.concurrent.FutureTask.run (FutureTask.java:264)
2024-08-13T11:59:11.2968385Z     at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
2024-08-13T11:59:11.2969842Z     at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
2024-08-13T11:59:11.2970954Z     at java.lang.Thread.run (Thread.java:840)
2024-08-13T11:59:11.2972157Z [WARNING] Exception occurred on node !debug only! [MethodCall:134:69]p.getBaseType().equals(Session.class)
```

**Code Sample demonstrating the issue:**

```java
	private static AnnotatedMethodConfigurator<?> configureWebSocketUserSessionInterceptor(
			ProcessAnnotatedType<?> type, Class<? extends Annotation> annotation, boolean required) {
		var method = type.configureAnnotatedType().filterMethods(m -> m.isAnnotationPresent(annotation)).findAny().orElse(null);
		if (method == null) {
			if (required) {
				throw new IllegalArgumentException(
						"Web socket endpoint " + type.getAnnotatedType().getJavaClass().getName() +
						" must have a @" + annotation.getSimpleName() + " method!");
			}
			return null;
		}
		
		if (method.getAnnotated().getParameters().stream().noneMatch(p -> p.getBaseType().equals(Session.class))) {
			throw new IllegalArgumentException(
					"Web socket endpoint " + type.getAnnotatedType().getJavaClass().getName() +
					" @" + annotation.getSimpleName() + " must have a Session parameter!");
		}
		
		method.add(new WebSocketUserSessionLiteral());
		return method;
	}
```

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1.  mvn clean package

**Running PMD through:**

Maven

<!-- If relevant, also include your JDK and OS information, e.g. for ClassNotFoundException, LinkageError, reflection failures, etc. -->

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeOps.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeOps.java
@@ -1170,6 +1170,13 @@ public final class TypeOps {
                     } else if (!upwards) {
                         // If Ai is a type that mentions a restricted type variable, then Ai' is undefined.
                         return NO_DOWN_PROJECTION;
+                    } else if (u instanceof JWildcardType) {
+                        // The rest of this function, below, treats u as the bound of a wildcard,
+                        // but if u is already a wildcard (and therefore ai was a wildcard), we
+                        // are already done.
+                        newTargs.add(u);
+                        change = true;
+                        continue;
                     }
 
                     change = true;

--- a/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/TypeOpsTest.kt
+++ b/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/types/TypeOpsTest.kt
@@ -95,7 +95,44 @@ class TypeOpsTest : FunSpec({
 
 
             }
+            test("#5167 problem in projection") {
+                val (acu, spy) = javaParser.parseWithTypeInferenceSpy(
+                    """
+import java.lang.annotation.Annotation;
+interface Bar<T> {
+    Baz<T> getBaz();
+}
+
+interface Predicate<T> {
+    boolean check(T t);
+}
+interface Stream<T>{
+    T findSome();
+}
+interface Baz<T>{
+    Stream<Bar<? super T>> filterMethods(Predicate<? super T> p);
+}
+
+class Foo {
+
+    private static Bar<?> foo(
+        Bar<?> type, Class<? extends Annotation> annotation, boolean required) {
+        var method = type.getBaz().filterMethods(m -> true).findSome();
+        return method;
+    }
+}
+            """.trimIndent()
+                )
+
+                val (barT) = acu.declaredTypeSignatures()
+                val methodId = acu.varId("method")
+
+                spy.shouldBeOk {
+                    methodId shouldHaveType barT[`?`]
+                }
+            }
         }
     }
+
 
 })

